assorted fixes for bundles

xeno-by opened the pull request:
This commit clearly establishes what bundles are. Previously it was
enough to just extend scala.reflect.macros.Macro, which created some
loopholes, but now scalac enforces that bundles:
  1) Are static (not necessarily top-level, but just static)
  2) Are traits (objects shouldn't be bundles anyway, and classes bring
     complications with their ctors which require special treatment in
     generated classes, so why support them if they don't bring anything
     new to the table?)
  3) Are monomorphic (again, this brings unnecessary complications wrt
     auxiliary code generation, so I don't see merit in supporting
     polymorphic bundles, whatever that a polymorphic bundle could mean)
  4) Don't provide concrete implementation for Macro.c (if they do then
     what is the point?)

Also fixes a couple issues. Firstly, bundles can now be nested as long
as they are static (which means that we can finally use bundles in REPL!).
Secondly, auxiliary code that's generated by the compiler alongside bundles
is now put into virtual compilation units with stable names, which solves
the infinite compilation problem in SBT 0.13.

xeno-by commented:
review @retronym @gkossakowski @harrah

retronym commented on src/compiler/scala/reflect/macros/compiler/Resolvers.scala:
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
-        val existingInvoker = rootMirror.getClassIfDefined(invokerFullName)
-        if (!currentRun.compiles(existingInvoker)) {
+        // synthesize the bundle, i.e. given a top-level `trait Foo extends Macro { def expand = ... } `
+        // create a top-level definition `class FooBundle(val c: Context) extends Foo` in MACRO_BUNDLE_PACKAGE
```
Is "top-level" in the comments still accurate? Or are things more general now ("static")?

xeno-by replied:
The first line should say "static", the second line should remain "top-level".

retronym commented on src/compiler/scala/reflect/macros/compiler/Resolvers.scala:
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
-        gen.mkTypeApply(Select(instanceOfInvoker, methName), targs)
+        // `new FooBundle(???).macroName` plus the optional type arguments
+        val bundleInstance = New(Select(bundlePid, bundleName), List(List(Select(scalaDot(nme.Predef), nme.???))))
+        gen.mkTypeApply(Select(bundleInstance, methName), targs)
```
Maybe I'm missing a later call to `atPos`, but this tree looks unpositioned.

xeno-by replied:
You're right

retronym commented on src/compiler/scala/reflect/macros/compiler/Resolvers.scala:
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
-        val instanceOfInvoker = New(Select(invokerPid, invokerName), List(List(Select(scalaDot(nme.Predef), nme.???))))
-        gen.mkTypeApply(Select(instanceOfInvoker, methName), targs)
+        // `new FooBundle(???).macroName` plus the optional type arguments
+        val bundleInstance = New(Select(bundlePid, bundleName), List(List(Select(scalaDot(nme.Predef), nme.???))))
```
It's not part of this patch, but we have `definitions.Predef_???` that could be used here.

xeno-by replied:
Thanks

retronym commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -311,9 +311,8 @@ trait StdNames {
-    // TODO: if I use dollars in MACRO_INVOKER_SUFFIX, as in "$Invoker$", then Scala reflection fails to load implementations
-    val MACRO_INVOKER_SUFFIX: NameType     = "Invoker"
+    val MACRO_BUNDLE_PACKAGE: NameType     = "scala.reflect.macros.bundles"
+    val MACRO_BUNDLE_SUFFIX: NameType      = "Bundle"
```
Perhaps `$Bundle` would make it a bit clearer that this is a compiler generated class.

xeno-by replied:
Done

retronym commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -311,9 +311,8 @@ trait StdNames {
-    val MACRO_INVOKER_PACKAGE: NameType    = "scala.reflect.macros.synthetic"
-    // TODO: if I use dollars in MACRO_INVOKER_SUFFIX, as in "$Invoker$", then Scala reflection fails to load implementations
-    val MACRO_INVOKER_SUFFIX: NameType     = "Invoker"
+    val MACRO_BUNDLE_PACKAGE: NameType     = "scala.reflect.macros.bundles"
```
Why not in the user's package of the bundle trait? I'm not comfortable with the compiler generating third party code underneath a global namespace, let alone one underneath `scala._`. Is there potential for clashes when mixing JARs defining bundles?

xeno-by replied:
You're right. I'll move synthetic code to the user's package.

retronym commented on src/compiler/scala/reflect/macros/compiler/Resolvers.scala:
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
+        // synthesize the bundle, i.e. given a top-level `trait Foo extends Macro { def expand = ... } `
+        // create a top-level definition `class FooBundle(val c: Context) extends Foo` in MACRO_BUNDLE_PACKAGE
+        val bundlePid = gen.mkUnattributedRef(nme.MACRO_BUNDLE_PACKAGE)
+        val bundleName = TypeName(bundleProto.fullName.split('.').map(_.capitalize).mkString("Dot") + nme.MACRO_BUNDLE_SUFFIX)
```
Okay, I see you use fully qualified names here. It would be more standard to encode them as `com$acme$Foo$Bundle` in the same way as the compiler does in `expandName` rather than inventing a new means to do this.

In any case, I'm still against putting them in this package, but I'm keen to hear what constraints led to this design.

(Sorry that I'm re-reviewing this old code now, I must have missed it when it went on originally.)

xeno-by replied:
Okay

xeno-by commented:
@retronym All feedback is addressed, and an SBT blocker is now fixed. Would be great to get a go for M6.

xeno-by committed:
clearly establishes what macro bundles are

Previously it was enough to just extend scala.reflect.macros.Macro, which
created some loopholes, but now scalac enforces that bundles:
  1) Are static (not necessarily top-level, but just static)
  2) Are traits (objects shouldn't be bundles anyway, and classes bring
     complications with their ctors which require special treatment in
     generated classes, so why support them if they don't bring anything
     new to the table?)
  3) Are monomorphic (again, this brings unnecessary complications wrt
     auxiliary code generation, so I don't see merit in supporting
     polymorphic bundles, whatever that a polymorphic bundle could mean)
  4) Don't provide concrete implementation for Macro.c (if they do then
     what is the point?)

xeno-by committed:
gets rid of randomness in virtual filenames for bundles

Bundles are emitted in compilation units that wrap virtual source files.
Previously we had those virtual files named randomly to ensure freshness,
but that led to infinite compilation loops in SBT (see the commit message
for more details and a link to a scala-user discussion).

Now the names are generated deterministically from full names of bundles,
which fixes the problems with SBT.

xeno-by committed:
macro bundles are now usable in repl

One of the previous commits relaxed the top-level restriction for bundles,
turning it into a requirement of staticness (i.e. bundles nested in static
objects are also okay now).

This means that we can now define bundles in repl. Almost.

There's still a little problem remaining that arises from the fact that
when compiling a line of input, repl doesn't automatically import all
previously defined symbols, but rather uses an heuristic to scan the
input and guess what symbols need to be imported.

Unfortunately for bundles, this heuristic fails, because when scanning
a macro definition that looks like `def foo = macro Macros.foo`, it thinks
that it's only necessary to import a term symbol called Macros (a vanilla
way of defining macro impls), but not a type symbol called Macros (a new
way of writing macro impls in bundles).

This commit fixes the problem by making the repl look for both term and
type symbols corresponding to the identifiers used in macro definitions.

xeno-by commented:
Rebased and split into more clear commits.

retronym commented:
LGTM. Thanks for addressing all of my feedback so quickly!
